presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java

```diff
@@ -0,0 +1,217 @@
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.DoubleType.DOUBLE;
+import static com.facebook.presto.common.type.IntegerType.INTEGER;
+import static com.facebook.presto.common.type.RealType.REAL;
+import static com.facebook.presto.type.TypeUtils.expectedValueSize;
+import static java.lang.Double.longBitsToDouble;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class KeySetAndValues
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(KeySetAndValues.class).instanceSize();
+    private static final int EXPECTED_ENTRIES = 10;
+    private static final int EXPECTED_ENTRY_SIZE = 16;
+
+    private final TypedSet keySet;
+    private final Type keyType;
+
+    private final BlockBuilder valueBlockBuilder;
+    private final Type valueType;
+
+    public KeySetAndValues(Type keyType, Type valueType)
+    {
+        this(keyType,
+                valueType,
+                new TypedSet(keyType, EXPECTED_ENTRIES, "MAP_UNION_SUM"),
+                valueType.createBlockBuilder(null, EXPECTED_ENTRIES, expectedValueSize(valueType, EXPECTED_ENTRY_SIZE)));
+    }
+
+    KeySetAndValues(Type keyType, Type valueType, TypedSet keySet, BlockBuilder valueBlockBuilder)
+    {
+        this.keyType = requireNonNull(keyType, "keyType is null");
+        this.valueType = requireNonNull(valueType, "valueType is null");
+        this.keySet = keySet;
+        this.valueBlockBuilder = valueBlockBuilder;
+    }
+
+    public Type getKeyType()
+    {
+        return keyType;
+    }
+
+    public Block getKeys()
+    {
+        return keySet.getBlock();
+    }
+
+    public Block getValues()
+    {
+        return valueBlockBuilder.build();
+    }
+
+    public void serialize(BlockBuilder out)
+    {
+        BlockBuilder mapBlockBuilder = out.beginBlockEntry();
+        Block keyBlock = keySet.getBlock();
+        for (int i = 0; i < keyBlock.getPositionCount(); i++) {
+            keyType.appendTo(keyBlock, i, mapBlockBuilder);
+            valueType.appendTo(valueBlockBuilder, i, mapBlockBuilder);
+        }
+        out.closeEntry();
+    }
+
+    public long estimatedInMemorySize()
+    {
+        return INSTANCE_SIZE + keySet.getRetainedSizeInBytes() + valueBlockBuilder.getRetainedSizeInBytes();
+    }
+
+    public void deserialize(Block block)
+    {
+        for (int i = 0; i < block.getPositionCount(); i += 2) {
+            keySet.add(block, i);
+            valueType.appendTo(block, i + 1, valueBlockBuilder);
+        }
+    }
+
+    /**
+     * Only add this key value pair if we haven't seen this key before.
+     * Otherwise, ignore it.
+     */
+    public KeySetAndValues unionSum(KeySetAndValues other)
+    {
+        Adder adder = getSummer(valueType);
+        TypedSet resultKeySet = new TypedSet(keyType, max(keySet.size(), other.keySet.size()), "MAP_UNION_SUM");
```

Comment:
Is it necessary to create the `resultKeySet` always? I thought you included the TypedSet into the aggregation state because you wanted to save some cost of building the hash tables. It should work to use the existing `keySet` member and add the elements in `other` to it because this is union semantic and no elements need to be deleted from existing `keySet`. The algorithm is something like this:
```
int[] matchedPositions = new int[valueBlockBuilder.getPositionCount()];
fill matchedPositions with -1;

for each element of position i in other {
    check if keySet contains it. If yes {
        matchedPositions[position in keySet] = i;
    }
    else {
        insert the current element to this.keySet. Note we have calculated the hash position already when testing contains.
        append this element to this.valueBlockBuilder;
     }
}

BlockBuilder resultValueBlockBuilder = valueType.createBlockBuilder(null, EXPECTED_ENTRIES, expectedValueSize(valueType, EXPECTED_ENTRY_SIZE));
for each position i in this.valueBlockBuilder, 
        if (matchedPositions[i] != -1) {
        append the sum (valueBlockBuilder.get(i) + otherValueBlock.get(matchedPositions[i]) to the resultValueBlockBuilder;
    }
    else {
        append the current value in this.valueBlockBuilder to resultValueBlockBuilder
    }
}
return new KeySetAndValues(keyType, valueType, keySet, resultValueBlockBuilder);
```
This way we can save some memory for creating a whole new KeySetAndValues, and should save the cost of adding the elements in `keySet` to `resultKeySet` which requires to compute hash positions again.

Reply:
Yeah - see the latest code - it's totally different. In fact, I'm refactoring it so we will only one set/group